rfcs/shared/build-system/ssr-testing.md

```diff
@@ -0,0 +1,240 @@
+
+- CommonJS is needed to avoid evaluation with `ts-node` or `babel-register`
+- ESM will be used directly in a browser
+
+The proposal is to use esbuild as it's blazing fast and perfectly matches this use case.
+
+<details>
+  <summary>`build.js`</summary>
+
+```ts
+await esbuild.build({
+  entryPoints: ['stories.tsx'],
+  bundle: true,
+  jsx: 'transform',
+  format: 'cjs',
+  target: 'node14',
+  // React & dependencies are not bundled intentionally to avoid collisions when they will be required
+  external: ['react', 'scheduler', 'react-dom', '@griffel/react', '@griffel/core'],
+  outfile: './dist/out-node.js',
+  plugins: [TsconfigPathsPlugin({ tsconfig: path.resolve(__dirname, 'tsconfig.base.json') })],
+});
+await esbuild.build({
+  // ESM has different entrypoint as we need to call "ReactDOM.hydrate()" there
+  entryPoints: ['app.tsx'],
+  bundle: true,
+  jsx: 'transform',
+  format: 'iife',
+  target: 'chrome101',
+  outfile: './dist/out-browser.js',
+  plugins: [TsconfigPathsPlugin({ tsconfig: path.resolve(__dirname, 'tsconfig.base.json') })],
+});
+```
+
+</details>
+
+**Note:** ESBuild will use existing path aliases to resolve packages used in stories.
+
+### Prepare assets
+
+On this stage we effectively need to create `index.html` that will be used for browser testing.
+
+We will use APIs for React 16.8/17 as we don't support React 18 yet. Once we will get React 18 in our repo it's reasonable to switch to streaming rendering. The snippet below renders all components and styles for them to static markup and saves as `index.html`.
+
+<details>
+  <summary>`render.js`</summary>
+
+```js
+const App = require('./out-node').App;
+const renderer = createDOMRenderer();
+
+const resultDOM = ReactDOM.renderToString(
+  React.createElement(RendererProvider, { renderer }, React.createElement(App)),
+);
+const resultStyles = ReactDOM.renderToStaticMarkup(renderToStyleElements(renderer));
+
+await fs.promises.writeFile(
+  './index.html',
+  `
+  <html>
+    <head>${resultStyles}</head>
+    <body>
+      <div id="root">${resultDOM}</div>
+      <script src="./out-browser.js"></script>
+    </body>
+  </html>
+  
+  `,
+);
+```
+
+</details>
+
+### Test
+
+This stage performs actual testing.
+
+- Start Chrome (via `puppeteer`)
+  - **Note**: the latest version of Puppeteer will be used to get latest Chrome. Existing version in our repo does not support even `:focus-visible` (that creates false positive errors)
+- Open a page (`index.html` file)
+  - **Note**: Webserver will not be used, a file will be opened via `file://`
+  - A page will have configured interceptors to avoid loading resources (images) from CDN
+- Ensure that there are no errors, otherwise test fails
+  - **Note**: For this reason the test app will run with `process.NODE_ENV = 'development'`
+  - **Note**: _Probably_ there is sense to restrict the test to catch only SSR related errors (in prototype it fails also on missing React keys, for example)
```

Comment:
I see why we might want to restrict to only SSR errors but it might be better to start by casting a wider net and fail on any error. Sure, the React keys example doesn't have anything to do with SSR but the fact that our story triggered the warning means the story wasn't demonstrating best practices so, imo, it's good that it failed even though that's not the explicit point of this test suite.

If we see lots of errors unrelated to SSR failing this test suite then I'd be in favor or restricting our failure modes.

Reply:
> I see why we might want to restrict to only SSR errors but it might be better to start by casting a wider net and fail on any error. Sure, the React keys example doesn't have anything to do with SSR but the fact that our story triggered the warning means the story wasn't demonstrating best practices so, imo, it's good that it failed even though that's not the explicit point of this test suite.

While I agree with this, I don't see that it's a goal of this test. As @Hotell it should be covered by a different tooling.